### PR TITLE
added solution for homework ArrayList

### PR DESCRIPTION
**Pull request by Valdes2**

### CODE REVIEW

File: src/main/java/core/basesyntax/ArrayList.java [hunk 1]
```diff
@@ -1,52 +1,108 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+
 /**
  * <p>Реалізувати свій ArrayList який імплементує інтерфейс List</p>
  */
 public class ArrayList<T> implements List<T> {
 
+    private int defaltCapacity = 10;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: make it as constant

File: src/main/java/core/basesyntax/ArrayList.java [hunk 2]
```diff
@@ -1,52 +1,108 @@
...
+    public Object[] elementData;
+    private int size = 0;
+
+    public ArrayList() {
+        elementData = new Object[defaltCapacity];
+    }
+
+    private void grow() {
+        if (size == elementData.length) {
+            elementData = Arrays.copyOf(elementData, elementData.length * 2);
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: why we increment size in twice?

File: src/main/java/core/basesyntax/ArrayList.java [hunk 3]
```diff
@@ -1,52 +1,108 @@
...
+        }
+        size = newSize;
+        elementData[size] = null;
+    }
+
     @Override
     public void add(T value) {
-
+        grow();
+        elementData[size] = value;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]:
```suggestion
        elementData[size++] = value;
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 4]
```diff
@@ -1,52 +1,108 @@
...
+        grow();
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: optional:
see how this method is implemented in the source code

File: src/main/java/core/basesyntax/ArrayList.java [hunk 5]
```diff
@@ -1,52 +1,108 @@
...
+    private void checkRange(int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayIndexOutOfBoundsException();
+        }
+    }
+
+    private void removeElement(Object[] element, int index) {
+        int newSize = size - 1;
+        if ((newSize) > index) {
+            System.arraycopy(elementData, index + 1, elementData, index,newSize - index);
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]:
```suggestion
            System.arraycopy(elementData, index + 1, elementData, index, newSize - index);
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 6]
```diff
@@ -1,52 +1,108 @@
...
+
+    private void checkRange(int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayIndexOutOfBoundsException();
+        }
+    }
+
+    private void removeElement(Object[] element, int index) {
+        int newSize = size - 1;
+        if ((newSize) > index) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: do we need this checking?

File: src/main/java/core/basesyntax/ArrayList.java [hunk 7]
```diff
@@ -1,52 +1,108 @@
...
     }
 
     @Override
     public T remove(T t) {
-        return null;
+        Object oldValue = null;
+        for (int i = 0; i < size; i++) {
+            if (elementData[i].equals(t)) {
+                oldValue = elementData[i];
+                removeElement(elementData, i);
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: add:
return oldValue;
and at the end of method we can remove `if` statement and write
            throw new ArrayIndexOutOfBoundsException();

File: src/main/java/core/basesyntax/ArrayList.java [hunk 8]
```diff
@@ -1,52 +1,108 @@
...
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return (size == 0);
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]:
```suggestion
        return size == 0;
```

File: src/main/java/core/basesyntax/ArrayList.java (hunk 1 above)
Comment (Valdes2) [reply src/main/java/core/basesyntax/ArrayList.java]: fixed

File: src/main/java/core/basesyntax/ArrayList.java (hunk 2 above)
Comment (Valdes2) [reply src/main/java/core/basesyntax/ArrayList.java]: Why not?)
Increment it in 1/2, like in source code?

File: src/main/java/core/basesyntax/ArrayList.java (hunk 3 above)
Comment (Valdes2) [reply src/main/java/core/basesyntax/ArrayList.java]: What diference between this and this:
```elementData[size] = value;
   size++;```

File: src/main/java/core/basesyntax/ArrayList.java (hunk 5 above)
Comment (Valdes2) [reply src/main/java/core/basesyntax/ArrayList.java]: fixed

File: src/main/java/core/basesyntax/ArrayList.java (hunk 8 above)
Comment (Valdes2) [reply src/main/java/core/basesyntax/ArrayList.java]: fixed

File: src/main/java/core/basesyntax/ArrayList.java (hunk 7 above)
Comment (Valdes2) [reply src/main/java/core/basesyntax/ArrayList.java]: okey, it's look like simpler

File: src/main/java/core/basesyntax/ArrayList.java [hunk 9]
```diff
@@ -1,52 +1,108 @@
...
+    }
+
+    private void checkRange(int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayIndexOutOfBoundsException();
+        }
+    }
+
+    private void removeElement(Object[] element, int index) {
+        int newSize = size - 1;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: index< size index <0 ?

File: src/main/java/core/basesyntax/ArrayList.java (hunk 6 above)
Comment (Valdes2) [reply src/main/java/core/basesyntax/ArrayList.java]: it`s necessary)
added checkIndexRange()

File: src/main/java/core/basesyntax/ArrayList.java (hunk 3 above)
Comment (Valdes2) [reply src/main/java/core/basesyntax/ArrayList.java]: fixed

File: src/main/java/core/basesyntax/ArrayList.java (hunk 2 above)
Comment (Valdes2) [reply src/main/java/core/basesyntax/ArrayList.java]: fixed

File: src/main/java/core/basesyntax/ArrayList.java (hunk 9 above)
Comment (Valdes2) [reply src/main/java/core/basesyntax/ArrayList.java]: fixed

**Pull request closed without merging**